File: ruby/sum-of-multiples/sum_of_multiples.rb
# frozen_string_literal: true

require 'set'

class SumOfMultiples
  def initialize(*factors)
    @factors = factors
  end

  def to(limit)
    multiples = Set.new
    @factors.each do |factor|
      next if factor.zero?

      multiple = factor
      while multiple < limit
        multiples << multiple
        multiple += factor
      end
    end

    multiples.sum
  end
end

File: java/sum-of-multiples/SumOfMultiples.java
import java.util.HashSet;

class SumOfMultiples {

    private int number;
    private int[] set;

    SumOfMultiples(int number, int[] set) {
        this.number = number;
        this.set = set;
    }

    int getSum() {
        throw new RuntimeException();
        HashSet<Integer> multiples = new HashSet<>();

        for (int factor: set) {
            if (factor == 0) {
                continue;
            }
            while (factor < number) {
                multiples.add(factor);
                factor += factor;
            }
        }
        return multiples.stream().mapToInt(Integer::intValue).sum();
    }

}
